crates/sui-network/src/randomness/mod.rs

```diff
@@ -0,0 +1,686 @@
+                let metrics = self.metrics.clone();
+                let authority_info = self.authority_info.clone();
+                let epoch = *epoch;
+                let round = *round;
+                let partial_sigs = ThresholdBls12381MinSig::partial_sign_batch(
+                    shares.iter(),
+                    &round.signature_message(),
+                );
+
+                // Record own partial sigs.
+                if !self.completed_sigs.contains(&(epoch, round)) {
+                    self.received_partial_sigs
+                        .insert((epoch, round, self.network.peer_id()), partial_sigs.clone());
+                    rounds_to_aggregate.push((epoch, round));
+                }
+
+                debug!("sending partial sigs for epoch {epoch}, round {round}");
+                spawn_monitored_task!(RandomnessEventLoop::send_partial_signatures_task(
+                    name,
+                    network,
+                    metrics,
+                    authority_info,
+                    epoch,
+                    round,
+                    partial_sigs
+                ))
+            });
+        }
+
+        if let Some(last_handled_key) = last_handled_key {
+            // Remove stuff from the pending_tasks map that we've handled.
+            let split_point = self
+                .pending_tasks
+                .range((Bound::Excluded(last_handled_key), Bound::Unbounded))
+                .next()
+                .cloned();
+            if let Some(key) = split_point {
+                self.pending_tasks = self.pending_tasks.split_off(&key);
+            } else {
+                self.pending_tasks.clear();
+            }
+        }
+        self.update_rounds_pending_metric();
+
+        // After starting a round, we have generated our own partial sigs. Check if that's
+        // enough for us to aggregate already.
+        for (epoch, round) in rounds_to_aggregate {
+            self.maybe_aggregate_partial_signatures(epoch, round);
+        }
+    }
+
+    async fn send_partial_signatures_task(
+        name: AuthorityName,
+        network: anemo::Network,
+        metrics: Metrics,
+        authority_info: Arc<HashMap<AuthorityName, (PeerId, PartyId)>>,
+        epoch: EpochId,
+        round: RandomnessRound,
+        partial_sigs: Vec<RandomnessPartialSignature>,
+    ) {
+        let _metrics_guard = metrics
+            .round_observation_latency_metric()
+            .map(|metric| metric.start_timer());
+
+        let peers: HashMap<_, _> = authority_info
+            .iter()
+            .map(|(name, (peer_id, _party_id))| (name, network.waiting_peer(*peer_id)))
+            .collect();
+        let partial_sigs: Vec<_> = partial_sigs
+            .iter()
+            .map(|sig| bcs::to_bytes(sig).expect("message serialization should not fail"))
+            .collect();
+
+        loop {
+            let mut requests = FuturesUnordered::new();
```

Comment:
a debug log before beginning the sending attempt could be helpful

Reply:
Anemo already has per-request logs built in